packages/shared/src/open-feature.ts

```diff
@@ -1,21 +1,208 @@
     return this._context;
   }
 
-  setTransactionContextPropagator(transactionContextPropagator: TransactionContextPropagator): OpenFeatureCommonAPI {
+  /**
+   * Adds a handler for the given provider event type.
+   * The handlers are called in the order they have been added.
+   * When changing the provider, the currently attached handlers will listen to the events of the new provider.
+   * @param {ProviderEvents} eventType The provider event type to listen to
+   * @param {EventHandler} handler The handler to run on occurrence of the event type
+   */
+  addHandler(eventType: ProviderEvents, handler: EventHandler): void {
+    this._events.addHandler(eventType, handler);
+  }
+
+  /**
+   * Removes a handler for the given provider event type.
+   * @param {ProviderEvents} eventType The provider event type to remove the listener for
+   * @param {EventHandler} handler The handler to remove for the provider event type
+   */
+  removeHandler(eventType: ProviderEvents, handler: EventHandler): void {
+    this._events.removeHandler(eventType, handler);
+  }
+
+  /**
+   * Gets the current handlers for the given provider event type.
+   * @param {ProviderEvents} eventType The provider event type to get the current handlers for
+   * @returns {EventHandler[]} The handlers currently attached to the given provider event type
+   */
+  getHandlers(eventType: ProviderEvents): EventHandler[] {
+    return this._events.getHandlers(eventType);
+  }
+
+  /**
+   * Sets the default provider for flag evaluations.
+   * This provider will be used by unnamed clients and named clients to which no provider is bound.
+   * Setting a provider supersedes the current provider used in new and existing clients without a name.
+   * @template P
+   * @param {P} provider The provider responsible for flag evaluations.
+   * @returns {OpenFeatureCommonAPI} OpenFeature API
+   */
+  setProvider(provider: P): this;
+  /**
+   * Sets the provider that OpenFeature will use for flag evaluations of providers with the given name.
+   * Setting a provider supersedes the current provider used in new and existing clients with that name.
+   * @template P
+   * @param {string} clientName The name to identify the client
+   * @param {P} provider The provider responsible for flag evaluations.
+   * @returns {this} OpenFeature API
+   */
+  setProvider(clientName: string, provider: P): this;
+  setProvider(clientOrProvider?: string | P, providerOrUndefined?: P): this {
+    const clientName = stringOrUndefined(clientOrProvider);
+    const provider = objectOrUndefined<P>(clientOrProvider) ?? objectOrUndefined<P>(providerOrUndefined);
+
+    if (!provider) {
+      return this;
+    }
+
+    const oldProvider = this.getProviderForClient(clientName);
+
+    // ignore no-ops
+    if (oldProvider === provider) {
+      return this;
+    }
+
+    if (clientName) {
+      this._clientProviders.set(clientName, provider);
+    } else {
+      this._defaultProvider = provider;
+    }
```

Comment:
I think there's one MINOR thing here that's from from the spec: https://github.com/open-feature/spec/blob/main/specification/sections/01-flag-evaluation.md#requirement-1122

> The provider mutator function MUST invoke the initialize function on the newly registered provider before using it to resolve flag values.

We should be initializing providers and THEN setting them, so that things "keep working" if the provider is switched.

Reply:
Makes sense, I will change it.